filters/include/bloom_filter_impl.hpp

```diff
@@ -0,0 +1,907 @@
+
+  ptr += copy_to_mem(num_hashes_, ptr);
+  ptr += copy_to_mem(static_cast<uint16_t>(0), ptr); // 2 bytes unused
+  ptr += copy_to_mem(seed_, ptr);
+  ptr += copy_to_mem(static_cast<int32_t>(capacity_bits_ >> 6), ptr); // sized in java longs
+  ptr += copy_to_mem(static_cast<uint32_t>(0), ptr); // 4 bytes unused
+
+  // rest of memory is num bits and bit array, so start with zeroes
+  std::fill_n(ptr, sizeof(uint64_t) * ((capacity_bits_ >> 6) + 1), 0);
+  bit_array_ = memory + BIT_ARRAY_OFFSET_BYTES;
+  memory_ = memory;
+}
+
+template<typename A>
+bloom_filter_alloc<A>::bloom_filter_alloc(const uint64_t seed,
+                                          const uint16_t num_hashes,
+                                          const bool is_dirty,
+                                          const bool is_owned,
+                                          const bool is_read_only,
+                                          const uint64_t capacity_bits,
+                                          const uint64_t num_bits_set,
+                                          uint8_t* bit_array,
+                                          uint8_t* memory,
+                                          const A& allocator) :
+  allocator_(allocator),
+  seed_(seed),
+  num_hashes_(num_hashes),
+  is_dirty_(is_dirty),
+  is_owned_(is_owned),
+  is_read_only_(is_read_only),
+  capacity_bits_((capacity_bits + 63) & ~0x3F),
+  num_bits_set_(num_bits_set),
+  bit_array_(bit_array),
+  memory_(memory)
+{
+  // private constructor
+  // no consistency checks since we should have done those prior to calling this
+  if (is_read_only_ && memory_ != nullptr && num_bits_set == DIRTY_BITS_VALUE) {
+    num_bits_set_ = bit_array_ops::count_num_bits_set(bit_array_, capacity_bits_ >> 3);
+  }
+}
+
+template<typename A>
+bloom_filter_alloc<A>::bloom_filter_alloc(const bloom_filter_alloc& other) :
+  allocator_(other.allocator_),
+  seed_(other.seed_),
+  num_hashes_(other.num_hashes_),
+  is_dirty_(other.is_dirty_),
+  is_owned_(other.is_owned_),
+  is_read_only_(other.is_read_only_),
+  capacity_bits_(other.capacity_bits_),
+  num_bits_set_(other.num_bits_set_)
+{
+  if (is_owned_) {
+    const size_t num_bytes = capacity_bits_ >> 3;
+    bit_array_ = AllocUint8(allocator_).allocate(num_bytes);
+    if (bit_array_ == nullptr) {
+      throw std::bad_alloc();
+    }
+    std::copy_n(other.bit_array_, num_bytes, bit_array_);
+    memory_ = nullptr;
+  } else {
+    bit_array_ = other.bit_array_;
+    memory_ = other.memory_;
+  }
+}
+
+template<typename A>
+bloom_filter_alloc<A>::bloom_filter_alloc(bloom_filter_alloc&& other) noexcept :
+  allocator_(std::move(other.allocator_)),
+  seed_(other.seed_),
+  num_hashes_(other.num_hashes_),
+  is_dirty_(other.is_dirty_),
+  is_owned_(other.is_owned_),
+  is_read_only_(other.is_read_only_),
+  capacity_bits_(other.capacity_bits_),
+  num_bits_set_(other.num_bits_set_),
+  bit_array_(std::move(other.bit_array_)),
```

Comment:
move doesn't make much sense for a pointer. this is just a copy

Reply:
yeah, i was trying to use copilot and while it helps with some boilerplate it inserts stuff like this, too, and for whatever reason i missed removing it.  thinking i need to turn that off overall.